Add backup attribute to file resource (close #124)

hico-horiuchi opened the pull request:
Add `backup` attribute to `file` (also `remote_file`, `template` and `http_request`).
The number of backups to be kept in `/var/chef/backup` (default value is `5`.)
(This attribute is same as [file](https://docs.chef.io/resource_file.html#properties) of Chef.)

For example:

```ruby
file `/tmp/file` do
  content 'Hello New World'
  backup 5
end

hico-horiuchi commented:
`/var/itamae/backup/tmp/file.itamae-20160114001759` is better ?

hico-horiuchi commented:
@ryotarai Following error occured in Werker at `start vm` step.
Could you confirm the error ?

```
There was an error when attemping to rsync a share folder.
Please inspect the error message below for more info.

Host path: /pipeline/build/spec/integration/
Guest path: /vagrant
Error: ssh: connect to host 104.131.69.157 port 22: Connection timed out
rsync: connection unexpectedly closed (0 bytes received so far) [sender]
rsync error: unexplained error (code 255) at io.c(605) [sender=3.0.9]
```

ryotarai commented on lib/itamae/resource/file.rb:
```diff
@@ -169,6 +176,34 @@ def send_tempfile
       end
+
+      def backup
+        return unless (attributes.backup).kind_of?(FalseClass) ? false : attributes.backup > 0
```
I think the following is simpler

```ruby
return if !attributes.backup || attributes.backup <= 0
```

hico-horiuchi replied:
Thank you! I'll fix it.

ryotarai commented on spec/integration/default_spec.rb:
```diff
@@ -51,9 +51,13 @@
+describe file('/var/itamae/backup/tmp') do
+  it { should be_directory }
+end
+
```
Would you add tests to check backed-up file?

hico-horiuchi replied:
I can't specify the backed-up file because file name is decided from date.
I try to use `commmand` resource with wildcard like below.

```ruby
describe command('cat /var/itamae/backup/tmp/file.itamae-*') do
  its(:stdout) { should match(/Hello World/) }
  its(:exit_status) { should eq(0) }
end
```

ryotarai replied:
> So, I try to use commmand resource with wildcard like below.

LGTM

hico-horiuchi commented on lib/itamae/resource/file.rb:
```diff
@@ -169,6 +176,34 @@ def send_tempfile
+        backup_files = run_command(['ls', '-1', backup_directory])
+        backup_files = backup_files.stdout.chomp.split("\n").select { |f|
+          f.match(/\A#{basename}.itamae-[0-9]+\z/)
+        }.reverse
```
Should I use `run_command("ls -1 #{::File.join(BACKUP_PATH, attributes.path + '.itamae-*')}")` ?
(When I use `['ls', '-1', ::File.join(BACKUP_PATH, attributes.path + '.itamae-*']` , `*` is escaped to `\*` and can't execute.)

ryotarai replied:
Current implementation is good.

hico-horiuchi replied:
Sure, thank you!

ryotarai commented on lib/itamae/resource/file.rb:
```diff
@@ -169,6 +176,34 @@ def send_tempfile
+        backup_path = ::File.join(BACKUP_PATH, backup_filename)
+        backup_directory = ::File.dirname(backup_path)
+
+        run_specinfra(:create_file_as_directory, backup_directory)
```
I have a concern about permission. For instance, the following situation can confuse users:

1. User A runs recipe A
    1. Recipe A creates a backup file of `/etc/foo.conf`
    1. `/var/itamae/backup/etc` is mode 755 and owned by user A
1. User B runs recipe B
    1. Recipe A creates a backup file of `/etc/bar.conf`
    1. User B can't create a directory or a file in `/var/itamae/backup/etc` because it is owned by user A

WDYT?

I think it is better all directories under `/var/lib/itamae` are `755`

hico-horiuchi replied:
@ryotarai Yes, I agree.

> I think it is better all directories under `/var/lib/itamae` are `755`

Isn't it mistake with `755` or `777` ?

ryotarai commented on lib/itamae/resource/file.rb:
```diff
@@ -169,6 +176,34 @@ def send_tempfile
+        backup_directory = ::File.dirname(backup_path)
+
+        run_specinfra(:create_file_as_directory, backup_directory)
+        run_specinfra(:copy_file, attributes.path, backup_path)
```
Files backed up should have the same permission. `cp -p` preserves the permission of source file.

hico-horiuchi replied:
`run_specinfra(:copy_file)` doesn't seem to preserve the permission.

  - [mizzy/specinfra/lib/specinfra/command/base/file.rb#L157#L159](https://github.com/mizzy/specinfra/blob/aa3678d56433a0b90da99810ef7a1d299659b041/lib/specinfra/command/base/file.rb#L157#L159)

So, I try to use `run_command(['cp', '-p', attributes.path, backup_path])` .

hico-horiuchi commented:
@ryotarai I apologize for my late reply.
I fixed following codes. Could you confirm?
(I also rebased commits.)

- Baked-up files have same permission of source files.
  - `run_command(['install', '-d', '-m=777', backup_directory])`
- Directories under `/var/lib/itamae` are `777` .
  - `run_command(['cp', '-p', attributes.path, backup_path])`

ryotarai commented:
Okay. It is better that `run_command` is replaced with `run_specinfra`.

hico-horiuchi commented:
@ryotarai `run_specinfra(:copy_file)` doesn't seem to preserve the permission.
So, I use `run_command(['cp', '-p', attributes.path, backup_path])` .

  - [mizzy/specinfra/lib/specinfra/command/base/file.rb#L157#L159](https://github.com/mizzy/specinfra/blob/3721727d3cd9e4b510ac88b21d581b99baa269b6/lib/specinfra/command/base/file.rb#L157#L159)

And `run_specinfra(:change_mode)` dosen't include `-R` option.
So, I use `run_command(['install', '-d', '-m=777', backup_directory])` .
(`run_specinfra(:create_file_as_directory, backup_directory)` + `run_command(['chmod', '-R', BACKUP_PATH])` is better?)

  - [mizzy/specinfra/lib/specinfra/command/base/file.rb#L140#L142](https://github.com/mizzy/specinfra/blob/3721727d3cd9e4b510ac88b21d581b99baa269b6/lib/specinfra/command/base/file.rb#L140#L142)

ryotarai commented:
As you said, Specinfra doesn't have sufficient commands for this PR but you can send a PR to https://github.com/mizzy/specinfra
Would you do it?

hico-horiuchi commented:
@ryotarai OK! I try to send PR to Specinfra.

hico-horiuchi commented:
@ryotarai I fixed by using Specinfra v2.54.0!
